### PR TITLE
Storybook docs

### PR DESCRIPTION
**Pull request by danielleflinn**
## Type of change
```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ X] Other
```

## Objective

Adding component documentation to Storybook. 

## Code changes

- **/clients/.storybook/preview.tsx:** updated sort order so components display alphabetically
- **/clients/libs/components/src/avatar/avatar.mdx:** added avatar component docs
- **/clients/libs/components/src/badge/badge.mdx:** added badge component docs
- **/clients/libs/components/src/breadcrumbs/breadcrumbs.mdx:** added breadcrumb component docs
- **/clients/libs/components/src/breadcrumbs/callout.mdx:** added callout component docs
- **/clients/libs/components/src/color-password/color-password.mdx:** added color-password component docs
- **/clients/libs/components/src/dialog/dialog/dialog.mdx:** added dialog component docs
- **/clients/libs/components/src/dialog/simple-dialog/simple-dialog.mdx:** added simple dialog component docs
- **/clients/libs/components/src/icon-button/icon-button.mdx:** added icon button component docs
- **/clients/libs/components/src/link/link.mdx:** added link component docs
- **/clients/libs/components/src/menu/menu.mdx:** added link component docs
- **/clients/libs/components/src/progress/progress.mdx:** added progress component docs
- **/clients/libs/components/src/table/table.mdx:** updated table component docs
- **/clients/libs/components/src/tabs/tabs.mdx:** added tabs component docs
- **/clients/libs/components/src/toggle-group/toggle-group.mdx:** added toggle-group component docs
- **/clients/libs/components/src/dialog/dialogs.mdx:** add general dialog page for main and simple dialog docs
- **/clients/libs/components/src/forms/forms.mdx:** updated form docs

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/components/src/button/button.mdx [hunk 1]
```diff
@@ -96,3 +71,39 @@ Typically button widths expand with their text. In some causes though buttons ma
...
+
+All button styles are WCAG compliant when displayed on `background` and `background-alt` colors. To
+use a button on a different background, double check that the color contrast is sufficient in both
+the light and dark themes.
+
+### Loading Buttons
+
+Include an `aria-label` attribute that defaults to “loading” but can be configurable per
+implementation. On click, the screen reader should announce the `aria-label`. Once the action is
+compelted, use another messaging pattern to alert the user that the action is complete (example:
```
Comment (coroiu) [inline libs/components/src/button/button.mdx:89]: typo
```suggestion
completed, use another messaging pattern to alert the user that the action is complete (example:
```

File: libs/components/src/button/button.mdx [hunk 2]
```diff
@@ -96,3 +71,39 @@ Typically button widths expand with their text. In some causes though buttons ma
...
+
+### Color contrast
+
+All button styles are WCAG compliant when displayed on `background` and `background-alt` colors. To
+use a button on a different background, double check that the color contrast is sufficient in both
+the light and dark themes.
+
+### Loading Buttons
+
+Include an `aria-label` attribute that defaults to “loading” but can be configurable per
```
Comment (coroiu) [inline libs/components/src/button/button.mdx:87]: suggestion: use the non-fancy quotation marks
```suggestion
Include an `aria-label` attribute that defaults to "loading" but can be configurable per
```

File: libs/components/src/button/button.mdx [hunk 3]
```diff
@@ -96,3 +71,39 @@ Typically button widths expand with their text. In some causes though buttons ma
...
+
+Both submit and async action buttons use a loading button state while an action is taken. If your
+button is preforming a long running task in the background like a server API call, be sure to review
+the [Async Actions Directive](?path=/story/component-library-async-actions-overview--page).
+
+<Story of={stories.Loading} />
+
+### appA11yTitle
+
+`appA11yTitle` is an element that auto assigns the same string to the `title` and `aria-label`
```
Comment (coroiu) [inline libs/components/src/button/button.mdx:102]: suggestion: match angular terminology
```suggestion
`appA11yTitle` is a directive that auto assigns the same string to the `title` and `aria-label`
```

File: libs/components/src/button/button.mdx [hunk 4]
```diff
@@ -96,3 +71,39 @@ Typically button widths expand with their text. In some causes though buttons ma
...
+attributes.
+
+When a button uses accessible content (i.e. actual text), DO NOT include this as it adds redundant
+content for someone using assistive technology.
+
+`appA11yTitle` should only be used if the element it applies to does not include accessible text,
+e.i. an icon.
```
Comment (coroiu) [inline libs/components/src/button/button.mdx:109]: typo:
```suggestion
i.e. an icon.
```

**Review by coroiu — COMMENTED** [review]
Looks good! Added some minor suggestions

File: libs/components/src/dialog/dialog/dialog.mdx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+### Large
+
+Use the large size for dialogs that have many interactive elements or tabbed content.
+
+**Tailwind styling:**
+
+` w-3/4` width: 75%;
+
+`max-w-3xl` 768px
```
Comment (Hinton) [inline libs/components/src/dialog/dialog/dialog.mdx:45]: This seems a bit confusing.

```
      case "small": {
        return "tw-max-w-sm";
      }
      case "large": {
        return "tw-max-w-3xl";
      }
      default: {
        return "tw-max-w-xl";
      }
```

File: libs/components/src/dialog/simple-dialog/simple-dialog.mdx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+icon: bwi-exclamation-triangle,
+color: warning,
+}
+
+if type = danger
+then {
+icon: bwi-error,
+color: danger,
+}
+```
```
Comment (Hinton) [inline libs/components/src/dialog/simple-dialog/simple-dialog.mdx:64]: I think a table would be more appropriate here.

File: libs/components/src/icon-button/icon-button.mdx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+indicator does not meet WCAG graphic contrast guidelines.
+
+## Sizes
+
+There are 2 sizes for the icon button: `small` and `default`.
+
+The padding on the `small` size uses .25rem on all sides, while the padding on the `default` size
+matches that of the button component.
+
+Both sizes display the icon at 1rem.
```
Comment (Hinton) [inline libs/components/src/icon-button/icon-button.mdx:38]: Do we need to include this in the storybook?

File: libs/components/src/link/link.mdx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+Text Links and Buttons use the `primary-500` color and can use either the `<a>` or `<button>` tags.
+Choose which based on the action the button takes:
+
+- if navigating to a new page, use a `<a>`
+- if taking an action on the current page use a `<button>`
+
+Text buttons or links are most commonly used in paragraphs of text or in forms to customize actions
+or show/hide additional form options.
+
+For focus states, use `focus-visible`.
```
Comment (Hinton) [inline libs/components/src/link/link.mdx:18]: Omit?

File: libs/components/src/link/link.mdx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+<Primary />
+<Controls />
+
+## Sizes
+
+There are 2 sizes for the component: default and small.
+
+Default uses `text-bas`e`and small uses`text-xs`
```
Comment (Hinton) [inline libs/components/src/link/link.mdx:27]:
```suggestion
Default uses `text-base`and small uses`text-xs`
```

Comment (Hinton) [reply libs/components/src/link/link.mdx:27]: I don't think we support size.

File: libs/components/src/menu/menu.mdx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+Menus are used to help organize related options. Menus are currently used for item options in
+tables, the profile menu to navigate to settings/log out, and is the base component for the Product
+Switcher.
+
+<Story of={stories.ClosedMenu} />
+<Controls />
+
+## Accessibility
+
+Follow WCAG AA best practices. Example.Insure the triggering element has `aria-haspopup="true"`
```
Comment (Hinton) [inline libs/components/src/menu/menu.mdx:18]:
```suggestion
Follow WCAG AA best practices. Example: Insure the triggering element has `aria-haspopup="true"`
```

**Review by Hinton — COMMENTED** [review]
Gone through part of it.

File: libs/components/src/avatar/avatar.mdx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+`textColor:` set dynamically based on background color to maintain a 4.5:1 contrast ratio.
+
+`backgroundColor:` defined by control, see next section
+
+`fontSize: 20px` since tailwind classes don’t render in the image function use 20px for the default
+size. Font-size will scale automatically from this.
+
+`fontWeight: 400`
+
+`charCount: 2`
```
[the remark this answers is not in the document]
Comment (danielleflinn) [reply libs/components/src/avatar/avatar.mdx:42]: Good point, we can probably cut this.

File: libs/components/src/avatar/avatar.mdx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+## Avatar as a button
+
+The Avatar can be used as a button.
+
+Typically this is only in the navigation on client apps where account switching is used and in the
+web app for the account menu indicator.
+
+When the avatar is used as a button, the following states should be used:
+
+`TODO: add stories for button avatars. See Figma`
```
[the remark this answers is not in the document]
Comment (jtouchstonebw) [reply libs/components/src/avatar/avatar.mdx:70]: I think it's a smart move to include current gaps in design system documentation. I argue we keep this, maybe add the link to Jira so we know it's already written?

[the remark this answers is not in the document]
Comment (danielleflinn) [reply libs/components/src/avatar/avatar.mdx:70]: I agree. I'll leave a note in docs pointing to the jira issue and figma.

File: libs/components/src/callout/callout.mdx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+## Styles
+
+Icons should remain consistent across these types. Do not change the icon without consulting
+designer. Use the following guidelines to help choose the correct type of callout.
+
+### Success
+
+Use the success callout to communicate a positive messaging to the user.
+
+**Example:** a positive report results shows a success callout.
```
[the remark this answers is not in the document]
Comment (danielleflinn) [reply libs/components/src/callout/callout.mdx:22]: We could. For now let's leave it to reduce the scope of this PR to just mdx changes.

File: libs/components/src/dialog/dialog/dialog.mdx (hunk 5 above)
Comment (danielleflinn) [reply libs/components/src/dialog/dialog/dialog.mdx:45]: Do we no longer set the width for dialogs? I thought we had both width and max-width set?

Comment (Hinton) [reply libs/components/src/dialog/dialog/dialog.mdx:34]: No, we don't set width. Which will support responsive designs later on.

File: libs/components/src/callout/callout.mdx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+# Callouts
+
+Callouts are used to communicate important information to the user. Callouts should be used
+sparingly, as they command a large amount of visual attention. Avoid using more than 1 callout in
+the same location.
+
+## Styles
+
+Icons should remain consistent across these types. Do not change the icon without consulting
+designer. Use the following guidelines to help choose the correct type of callout.
```
Comment (Hinton) [inline libs/components/src/callout/callout.mdx:16]: Feels like we're missing a `the`?

File: libs/components/src/menu/menu.mdx [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+import * as stories from "./menu.stories";
+
+<Meta of={stories} />
+
+# Menu
+
+Menus are used to help organize related options. Menus are currently used for item options in
+tables, the profile menu to navigate to settings/log out, and is the base component for the Product
+Switcher.
```
Comment (Hinton) [inline libs/components/src/menu/menu.mdx:11]: I wonder if we should rephrase this to "currently used for things like item options", otherwise we should keep track of new usages and add it to the list.

File: libs/components/src/progress/progress.mdx [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+<Story of={stories.Full} />
+
+## Text label
+
+When measuring something other than progress, such as password strength, update the label to fit the
+context of the implementation.
+
+<Story of={stories.CustomText} />
+&nbsp;{" "}
```
Comment (Hinton) [inline libs/components/src/progress/progress.mdx:30]: And reason to have this?

File: libs/components/src/progress/progress.mdx [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+- **Good:** `primary-500`
+- **Strong:** `success-500`
+
+## Accessibility
+
+Be sure to include the proper `aria-valuemin`, `aria-valuemax`, and `aria-valuenow` attributes. An
+a`ria-valuetext` should also be configurable to include the text a screen reader should read to the
+user.
+
+In the case of a password strength indicator; `aria-describedb`y should be used and should point to
```
Comment (Hinton) [inline libs/components/src/progress/progress.mdx:47]:
```suggestion
In the case of a password strength indicator; `aria-describedby` should be used and should point to
```

Comment (Hinton) [reply libs/components/src/progress/progress.mdx:47]: Should we rewrite this to be in the present tense, essentially describe that the component uses instead of should.

File: libs/components/src/stories/dialogs.mdx
Comment (Hinton) [inline libs/components/src/stories/dialogs.mdx:1]: We should move this file to `src/dialogs/dialogs.mdx`.

File: libs/components/src/stories/dialogs.mdx [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+A user should not be able to tab focus outside of the Dialog until it has been closed.
+
+Once closed, focus should remain on the the element which triggered the Dialog.
+
+Clicking outside the dialog or clicking escape should close the dialog (this prevents a keyboard
+trap)
+
+**Note:** If a Simple Dialog is triggered from a main Dialog, be sure to make sure focus is moved to
+the Simple Dialog.
```
Comment (Hinton) [inline libs/components/src/stories/dialogs.mdx:55]: We should go over this and identify what the consumer needs to implement on their side, and what the component already provides.

File: libs/components/src/stories/forms.mdx [hunk 19]
```diff
@@ -4,48 +4,175 @@ import { Meta, Story, Source } from "@storybook/addon-docs";
...
+
+<Source id="component-library-form--full-example" />
+
+## Form spacing and sections
+
+Forms consists of 1 or more inputs, and ends with 1 or 2 buttons.
 
-### Form Field
+If there are many inputs in a form, they should should be organized into sections as content
+relates. **Example:** Item type form
```
Comment (Hinton) [inline libs/components/src/stories/forms.mdx:20]: Might be nice with a screenshot of this form.

File: libs/components/src/stories/forms.mdx [hunk 20]
```diff
@@ -4,48 +4,175 @@ import { Meta, Story, Source } from "@storybook/addon-docs";
...
+button itself.
+
+If a checkbox group has more than 4 options a
+[multi-select components](?path=/docs/component-library-form-multi-select--docs) should be used.
+
+#### Single checkbox
+
+<Story id="component-library-form-checkbox--default" />
+
+## Accessibility
```
Comment (Hinton) [inline libs/components/src/stories/forms.mdx:121]: Same here, would be beneficial to know what you are expected to do yourself vs what is automatically done by the CL.

File: libs/components/src/tabs/tabs.mdx [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+The tab navigation and content tabs share the same styling. The tab navigation uses links to
+navigate between pages, whereas the tab list uses `<buttons>` to toggle content on a single page.
+
+Tabs should be displayed on the `background-alt` color, with their content area set to background
+and 1rem of padding on the left and right.
+
+~20px of padding is used bellow the tabs and before the section content.
+
+For focus states, use `focus-visible`.
```
Comment (Hinton) [inline libs/components/src/tabs/tabs.mdx:17]: I think we can remove this.

File: libs/components/src/tabs/tabs.mdx [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+## Content tabs in dialogs
+
+Tabs can be used in dialogs to separate related content.
+
+<Story id="component-library-dialogs-dialog--tab-content" />
+
+## Accessibility
+
+**Navigation tabs** should be implemented using the `<nav>` element and `<a>` for each tab.
```
Comment (Hinton) [inline libs/components/src/tabs/tabs.mdx:38]:
```suggestion
**Navigation tabs** are implemented using the `<nav>` element and `<a>` for each tab.
```

File: libs/components/src/progress/progress.mdx (hunk 16 above)
Comment (danielleflinn) [reply libs/components/src/progress/progress.mdx:30]: Yes, for some reason the h3 is not rendering with space above it without the `&nbsp;{" "}`
Screenshot 2023-06-26 at 10 19 28 AM

File: libs/components/src/stories/dialogs.mdx (hunk 18 above)
Comment (danielleflinn) [reply libs/components/src/stories/dialogs.mdx]: Updated in lates commit. It would be good if a Front end engineer could review this for accuracy.

File: libs/components/src/stories/forms.mdx (hunk 19 above)
Comment (danielleflinn) [reply libs/components/src/stories/forms.mdx]: I'd like to avoid adding screenshots since they will become difficult to maintain if the visual changes. Perhaps we add a embed the story of this form once it has been migrated to the library.

File: libs/components/src/stories/forms.mdx (hunk 20 above)
Comment (danielleflinn) [reply libs/components/src/stories/forms.mdx]: I'll add a jira issue to work on this as I'll need a front-end engineers assistance identifying which behavior is provided and which needs to be implemented per form.

File: libs/components/src/badge/badge.mdx [hunk 23]
```diff
@@ -23,18 +23,37 @@ The primary badge is used to indicate an active status (example: device manageme
...
 neutral information.
+
 <Story of={stories.Secondary} />
-### Success The success badge is used to indicate a positive status, OR to indicate a feature requires
-a Premium subscription. See [Premium Badge](?path=/docs/web-premium-badge--docs)
+
+### Success
+
+The success badge is used to indicate a positive status, OR to indicate a feature requires a Premium
+subscription. See [Premium Badge](?path=/docs/web-premium-badge--docs)
```
Comment (Hinton) [inline libs/components/src/badge/badge.mdx:37]: Should we just add the premium badge below success?

File: libs/components/src/dialog/dialogs.mdx [hunk 24]
```diff
@@ -28,12 +28,18 @@ A backdrop should be used to hide the content below the dialog. Use `#000000` wi
 
 ## Accessibility
 
-The triggering button should indicate to assistive technology that additional content will open or
-appear when the trigger is selected. Consider using `aria-haspopup="true"`
+### Component behavior
 
-Dialog should include `role=“dialog”`
+- Dialog include `role=“dialog”`
```
Comment (Hinton) [inline libs/components/src/dialog/dialogs.mdx:33]:
```suggestion
- Dialog include `role="dialog"`
```

**Review by Hinton — DISMISSED** [review]

**Review by Hinton — APPROVED** [review]
